src/System.CommandLine/Binding/ReflectionBinder.cs

```diff
@@ -0,0 +1,426 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.CommandLine.Invocation;
+using System.Linq;
+using System.Reflection;
+
+namespace System.CommandLine.Binding
+{
+    public class ReflectionBinder : IBinder
+    {
+        public ReflectionBinder(Type type)
+            => Type = type;
+
+        private const BindingFlags CommonBindingFlags = BindingFlags.FlattenHierarchy
+                                    | BindingFlags.IgnoreCase
+                                    | BindingFlags.Public
+                                    | BindingFlags.NonPublic;
+
+        private const BindingFlags IgnorePrivateBindingFlags = BindingFlags.FlattenHierarchy
+                                | BindingFlags.IgnoreCase
+                                | BindingFlags.Public
+                                | BindingFlags.NonPublic;
+
+        private object _explicitlySetTarget;
+        private Type Type { get; }
+        private  MethodInfo InvocationMethodInfo { get;  set; }
+        // I really hate the location of these. I think we need a reflection binding set that incorporates these
+        private ParameterCollection InvocationParameterCollection { get;  set; }
+        private ParameterCollection ConstructorParameterCollection { get;  set; }
+
+        private readonly BindingSet ConstructorBindingSet = new BindingSet();
+        private readonly BindingSet InvocationBindingSet = new BindingSet();
+        private bool isBoundToCommand;
```

Comment:
For consistency, re: the above declarations, please use `_` prefixes and make them fields if they are `private`.